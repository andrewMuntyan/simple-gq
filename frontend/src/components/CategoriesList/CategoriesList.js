import React from 'react';
import PropTypes from 'prop-types';

import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import { ListRow } from '../ListRow';
import { List } from '../List';

export const GET_CATEGORIES = gql`
  query GET_CATEGORIES {
    categories {
      id
      name
    }
  }
`;

const CategoriesList = ({ onCategoryPick, selectedCategory }) => {
  const { data, loading, error, fetchMore } = useQuery(GET_CATEGORIES);
  const { categories = [] } = data;

  if (loading) {
    return <h2>Loading Categories...</h2>;
  }

  // TODO: add spinner
  return categories.length ? (
    <List
      itemsData={categories}
      RowRenderer={ListRow}
      onItemClick={onCategoryPick}
      selectedItem={selectedCategory}
    />
  ) : (
    <h2>There is Nothing... Add new Category!</h2>
  );
};

CategoriesList.propTypes = {
  onCategoryPick: PropTypes.func.isRequired,
  selectedCategory: PropTypes.string
};
CategoriesList.defaultProps = {
  selectedCategory: null
};

export default CategoriesList;
