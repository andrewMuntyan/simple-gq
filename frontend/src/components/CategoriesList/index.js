export { default as CategoriesList } from './CategoriesList';
export { GET_CATEGORIES } from './CategoriesList';
