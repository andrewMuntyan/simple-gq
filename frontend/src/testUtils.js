import { act } from 'react-dom/test-utils';

export const fakeWord = () => ({
  content: 'word',
  id: 'uniqueId',
  createdAt: '2019-09-11T21:13:39.695Z'
});

export const getFakeCategoriesData = () => ({
  categories: [
    {
      id: 'uuid1',
      name: 'category 1',
      __typename: 'Category',
      createdAt: '',
      updatedAt: ''
    },
    {
      id: 'uuid2',
      name: 'category 2',
      __typename: 'Category',
      createdAt: '',
      updatedAt: ''
    },
    {
      id: 'uuid3',
      name: 'category 3',
      __typename: 'Category',
      createdAt: '',
      updatedAt: ''
    }
  ]
});

export const fakeCategory = () => getFakeCategoriesData().categories[0];

export const fakePaginationResponse = (count = 0) => ({
  data: {
    wordsConnection: {
      aggregate: { count, __typename: 'AggregateWord' },
      __typename: 'WordConnection'
    }
  }
});

// see https://github.com/airbnb/enzyme/issues/2073
export const waitAndUpdateWrapper = async (wrapper, time = 0) => {
  await act(async () => {
    await new Promise(res => setTimeout(res, time));
    await wrapper.update();
  });
};
