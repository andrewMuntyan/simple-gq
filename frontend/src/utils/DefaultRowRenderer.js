import React from 'react';
import PropTypes from 'prop-types';

const DefaultRowRenderer = ({ name }) => {
  return <div>{name}</div>;
};

DefaultRowRenderer.propTypes = {
  name: PropTypes.string.isRequired
};

export default DefaultRowRenderer;
