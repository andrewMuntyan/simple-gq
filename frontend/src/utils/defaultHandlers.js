export const defaultSubmit = e => {
  e.preventDefault();
  // eslint-disable-next-line no-console
  console.warn('Default Submit handler is triggered');
};

export const defaultOnChange = () => {
  // eslint-disable-next-line no-console
  console.warn('Default Change handler is triggered');
};
