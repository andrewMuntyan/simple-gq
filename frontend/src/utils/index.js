export { default as DefaultRowRenderer } from './DefaultRowRenderer';
