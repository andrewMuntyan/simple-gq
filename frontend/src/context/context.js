import React from 'react';
import PropTypes from 'prop-types';
import { SnackbarProvider, useSnackbar } from 'notistack';
import { ErrorBoundary } from '../components';

import { getMessage } from '../utils';

const defaultContextValue = {
  selectedCategory: '',
  searchInProgress: false,
  searchTerm: ''
};

export const AppStateCtx = React.createContext(defaultContextValue);
export const AppStateSetterCtx = React.createContext(null);
export const SnackBarCtx = React.createContext({});

// TODO: move `SnackBarProviderWrapper` and `SnackBarContextProvider` to separate file

const SnackBarProviderWrapper = ({ children }) => (
  <SnackbarProvider maxSnack={3}>
    <SnackBarContextProvider>{children}</SnackBarContextProvider>
  </SnackbarProvider>
);
SnackBarProviderWrapper.propTypes = {
  children: PropTypes.node // or PropTypes.node.isRequired to make it required
};
SnackBarProviderWrapper.defaultProps = {
  children: null
};

const SnackBarContextProvider = ({ children }) => {
  const { enqueueSnackbar } = useSnackbar();

  const showMessage = ({ variant, text }) => {
    const isError = variant === 'error';
    // variant could be success, error, warning, info, or default
    enqueueSnackbar(text, {
      variant,
      anchorOrigin: { vertical: 'bottom', horizontal: 'right' },
      autoHideDuration: isError ? 4000 : 2000
      // preventDuplicate: isError
    });
  };

  const onActionDone = result => showMessage(getMessage(result));
  return (
    <SnackBarCtx.Provider value={{ onActionDone }}>
      {children}
    </SnackBarCtx.Provider>
  );
};

SnackBarContextProvider.propTypes = {
  children: PropTypes.node // or PropTypes.node.isRequired to make it required
};
SnackBarContextProvider.defaultProps = {
  children: null
};

export class AppSharedStateProvider extends React.Component {
  constructor(props) {
    super(props);
    // const { enqueueSnackbar } = useSnackbar();
    this.state = {
      appState: { ...defaultContextValue }
    };
  }

  setAppState = (newValues = {}) => {
    this.setState(prevState => ({
      ...prevState,
      appState: { ...prevState.appState, ...newValues }
    }));
  };

  render() {
    const { appState } = this.state;
    const { children } = this.props;
    return (
      <AppStateCtx.Provider value={appState}>
        <AppStateSetterCtx.Provider value={this.setAppState}>
          {children}
        </AppStateSetterCtx.Provider>
      </AppStateCtx.Provider>
    );
  }
}
AppSharedStateProvider.propTypes = {
  children: PropTypes.node // or PropTypes.node.isRequired to make it required
};
AppSharedStateProvider.defaultProps = {
  children: null
};

export const AppProvider = ({ children }) => (
  <SnackBarProviderWrapper>
    <AppSharedStateProvider>
      <ErrorBoundary>{children}</ErrorBoundary>
    </AppSharedStateProvider>
  </SnackBarProviderWrapper>
);

AppProvider.propTypes = {
  children: PropTypes.node // or PropTypes.node.isRequired to make it required
};
AppProvider.defaultProps = {
  children: null
};
