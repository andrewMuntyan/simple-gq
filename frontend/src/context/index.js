import React, { useState } from 'react';
import { SnackbarProvider, useSnackbar } from 'notistack';
import { Shield } from '../components';

import { getMessage } from '../utils';

const defaultContextValue = {
  selectedCategory: ''
};

export const AppContext = React.createContext([defaultContextValue, () => {}]);

// TODO: add proptypes
// eslint-disable-next-line react/prop-types
export const AppContextProvider = ({ children }) => {
  const { enqueueSnackbar } = useSnackbar();

  const showMessage = ({ variant, text }) => {
    const isError = variant === 'error';
    // variant could be success, error, warning, info, or default
    enqueueSnackbar(text, {
      variant,
      anchorOrigin: { vertical: 'bottom', horizontal: 'right' },
      autoHideDuration: isError ? 4000 : 2000
      // preventDuplicate: isError
    });
  };

  const onActionDone = result => showMessage(getMessage(result));

  const [appContext, setAppContext] = useState({
    ...defaultContextValue,
    showMessage,
    onActionDone
  });

  return (
    <AppContext.Provider value={[appContext, setAppContext]}>
      {/* Error Boundary */}
      <Shield>{children}</Shield>
    </AppContext.Provider>
  );
};

export const FinalAppContextProvider = props => {
  return (
    <SnackbarProvider maxSnack={3}>
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <AppContextProvider {...props} />
    </SnackbarProvider>
  );
};
