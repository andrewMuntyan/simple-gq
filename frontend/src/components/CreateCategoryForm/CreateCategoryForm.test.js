import React from 'react';

import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';
import { InMemoryCache } from 'apollo-cache-inmemory';

import { CreateCategoryForm, CREATE_CATEGORY_MUTATION } from '.';
import { GET_CATEGORIES } from '../CategoriesList';
import { AppContext } from '../../context';
import { getMessage } from '../../utils';

import { fakeCategory } from '../../testUtils';

const categoryData = fakeCategory();
const { name: newCategoryName } = categoryData;
const mutationVariables = { name: newCategoryName };
const newCategoryResponseData = {
  createCategory: categoryData
};

const mocks = [
  {
    request: {
      query: CREATE_CATEGORY_MUTATION,
      variables: mutationVariables
    },
    result: jest.fn(() => ({ data: newCategoryResponseData }))
  }
];

describe('<CreateCategoryForm />', () => {
  it('renders and maches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateCategoryForm />
      </MockedProvider>
    );

    const formComponent = wrapper.find('[data-test="add-entity-form"]');
    expect(formComponent).toHaveLength(1);
    expect(toJSON(formComponent)).toMatchSnapshot();
  });

  it('calls the mutation and updates local cache', async () => {
    // since <CreateCategoryForm /> component interacts
    // with apollo local cashe after create Category mutation is made
    // we have to prepare sych local apollo cache
    const cache = new InMemoryCache();
    cache.writeQuery({
      query: GET_CATEGORIES,
      data: { categories: [] }
    });

    // cache initialised with empty array for categories
    const { categories: initialCategoriesData } = cache.readQuery({
      query: GET_CATEGORIES
    });
    expect(initialCategoriesData).toBeInstanceOf(Array);
    expect(initialCategoriesData).toHaveLength(0);

    const showMessage = jest.fn();
    const wrapper = mount(
      <MockedProvider mocks={mocks} cache={cache}>
        <AppContext.Provider value={[{ showMessage }]}>
          <CreateCategoryForm />
        </AppContext.Provider>
      </MockedProvider>
    );

    const inputSelector = '[data-test="text-field"]';
    const formSelector = '[data-test="add-entity-form"]';

    // we have the input control
    const pristineInput = wrapper.find(inputSelector);
    expect(pristineInput).toHaveLength(1);
    expect(pristineInput.props().value).toEqual('');

    // typing to input
    pristineInput.simulate('change', {
      target: { value: newCategoryName }
    });

    await wait();
    wrapper.update();

    // input value is changed
    const dirtyInput = wrapper.find(inputSelector);
    expect(dirtyInput.props().value).toEqual(newCategoryName);

    // form submitted
    const form = wrapper.find(formSelector);
    expect(form).toHaveLength(1);
    form.simulate('submit');

    await wait(50);

    // Check if mutation has been called
    expect(mocks[0].result).toHaveBeenCalledTimes(1);

    // Check if success message was shown
    expect(showMessage).toHaveBeenCalledTimes(1);
    expect(showMessage).toHaveBeenCalledWith(getMessage(true));

    // cache is updated with data from mocked response
    const { categories: resultedCategoriesData } = cache.readQuery({
      query: GET_CATEGORIES
    });

    expect(resultedCategoriesData).toBeInstanceOf(Array);
    expect(resultedCategoriesData).toHaveLength(1);
    expect(resultedCategoriesData[0]).toEqual(categoryData);
  });

  // TODO: add test for case whent there is an error
});
