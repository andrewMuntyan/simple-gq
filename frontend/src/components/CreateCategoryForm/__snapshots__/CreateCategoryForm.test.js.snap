// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateCategoryForm /> renders and maches snapshot 1`] = `
<form
  className="makeStyles-container-1"
  data-test="add-entity-form"
  noValidate={true}
  onSubmit={[Function]}
>
  <span
    className="makeStyles-input-2"
  >
    <TextInput
      disabled={false}
      error={false}
      label="Add new category"
      onChange={[Function]}
      required={true}
      value=""
    >
      <WithStyles(ForwardRef(TextField))
        autoComplete="off"
        className="makeStyles-textField-5"
        disabled={false}
        error={false}
        id="standard-dense"
        inputProps={
          Object {
            "data-test": "text-field",
          }
        }
        label="Add new category"
        margin="dense"
        onChange={[Function]}
        required={true}
        value=""
      >
        <ForwardRef(TextField)
          autoComplete="off"
          className="makeStyles-textField-5"
          classes={
            Object {
              "root": "MuiTextField-root",
            }
          }
          disabled={false}
          error={false}
          id="standard-dense"
          inputProps={
            Object {
              "data-test": "text-field",
            }
          }
          label="Add new category"
          margin="dense"
          onChange={[Function]}
          required={true}
          value=""
        >
          <WithStyles(ForwardRef(FormControl))
            className="MuiTextField-root makeStyles-textField-5"
            disabled={false}
            error={false}
            margin="dense"
            required={true}
            variant="standard"
          >
            <ForwardRef(FormControl)
              className="MuiTextField-root makeStyles-textField-5"
              classes={
                Object {
                  "fullWidth": "MuiFormControl-fullWidth",
                  "marginDense": "MuiFormControl-marginDense",
                  "marginNormal": "MuiFormControl-marginNormal",
                  "root": "MuiFormControl-root",
                }
              }
              disabled={false}
              error={false}
              margin="dense"
              required={true}
              variant="standard"
            >
              <div
                className="MuiFormControl-root MuiTextField-root makeStyles-textField-5 MuiFormControl-marginDense"
              >
                <WithStyles(ForwardRef(InputLabel))
                  htmlFor="standard-dense"
                >
                  <ForwardRef(InputLabel)
                    classes={
                      Object {
                        "animated": "MuiInputLabel-animated",
                        "asterisk": "MuiInputLabel-asterisk",
                        "disabled": "Mui-disabled",
                        "error": "Mui-error",
                        "filled": "MuiInputLabel-filled",
                        "focused": "Mui-focused",
                        "formControl": "MuiInputLabel-formControl",
                        "marginDense": "MuiInputLabel-marginDense",
                        "outlined": "MuiInputLabel-outlined",
                        "required": "Mui-required",
                        "root": "MuiInputLabel-root",
                        "shrink": "MuiInputLabel-shrink",
                      }
                    }
                    htmlFor="standard-dense"
                  >
                    <WithStyles(ForwardRef(FormLabel))
                      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
                      classes={
                        Object {
                          "asterisk": "MuiInputLabel-asterisk",
                          "disabled": "Mui-disabled",
                          "error": "Mui-error",
                          "focused": "Mui-focused",
                          "required": "Mui-required",
                        }
                      }
                      data-shrink={false}
                      htmlFor="standard-dense"
                    >
                      <ForwardRef(FormLabel)
                        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
                        classes={
                          Object {
                            "asterisk": "MuiFormLabel-asterisk MuiInputLabel-asterisk",
                            "disabled": "Mui-disabled Mui-disabled",
                            "error": "Mui-error Mui-error",
                            "filled": "MuiFormLabel-filled",
                            "focused": "Mui-focused Mui-focused",
                            "required": "Mui-required Mui-required",
                            "root": "MuiFormLabel-root",
                          }
                        }
                        data-shrink={false}
                        htmlFor="standard-dense"
                      >
                        <label
                          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-required Mui-required"
                          data-shrink={false}
                          htmlFor="standard-dense"
                        >
                          Add new category
                          <span
                            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                          >
                            â€‰
                            *
                          </span>
                        </label>
                      </ForwardRef(FormLabel)>
                    </WithStyles(ForwardRef(FormLabel))>
                  </ForwardRef(InputLabel)>
                </WithStyles(ForwardRef(InputLabel))>
                <WithStyles(ForwardRef(Input))
                  autoComplete="off"
                  id="standard-dense"
                  inputProps={
                    Object {
                      "data-test": "text-field",
                    }
                  }
                  onChange={[Function]}
                  value=""
                >
                  <ForwardRef(Input)
                    autoComplete="off"
                    classes={
                      Object {
                        "disabled": "Mui-disabled",
                        "error": "Mui-error",
                        "focused": "Mui-focused",
                        "formControl": "MuiInput-formControl",
                        "fullWidth": "MuiInput-fullWidth",
                        "input": "MuiInput-input",
                        "inputMarginDense": "MuiInput-inputMarginDense",
                        "inputMultiline": "MuiInput-inputMultiline",
                        "inputTypeSearch": "MuiInput-inputTypeSearch",
                        "multiline": "MuiInput-multiline",
                        "root": "MuiInput-root",
                        "underline": "MuiInput-underline",
                      }
                    }
                    id="standard-dense"
                    inputProps={
                      Object {
                        "data-test": "text-field",
                      }
                    }
                    onChange={[Function]}
                    value=""
                  >
                    <WithStyles(ForwardRef(InputBase))
                      autoComplete="off"
                      classes={
                        Object {
                          "disabled": "Mui-disabled",
                          "error": "Mui-error",
                          "focused": "Mui-focused",
                          "formControl": "MuiInput-formControl",
                          "fullWidth": "MuiInput-fullWidth",
                          "input": "MuiInput-input",
                          "inputMarginDense": "MuiInput-inputMarginDense",
                          "inputMultiline": "MuiInput-inputMultiline",
                          "inputTypeSearch": "MuiInput-inputTypeSearch",
                          "multiline": "MuiInput-multiline",
                          "root": "MuiInput-root MuiInput-underline",
                          "underline": null,
                        }
                      }
                      fullWidth={false}
                      id="standard-dense"
                      inputComponent="input"
                      inputProps={
                        Object {
                          "data-test": "text-field",
                        }
                      }
                      multiline={false}
                      onChange={[Function]}
                      type="text"
                      value=""
                    >
                      <ForwardRef(InputBase)
                        autoComplete="off"
                        classes={
                          Object {
                            "adornedEnd": "MuiInputBase-adornedEnd",
                            "adornedStart": "MuiInputBase-adornedStart",
                            "disabled": "Mui-disabled Mui-disabled",
                            "error": "Mui-error Mui-error",
                            "focused": "Mui-focused Mui-focused",
                            "formControl": "MuiInputBase-formControl MuiInput-formControl",
                            "fullWidth": "MuiInputBase-fullWidth MuiInput-fullWidth",
                            "input": "MuiInputBase-input MuiInput-input",
                            "inputAdornedEnd": "MuiInputBase-inputAdornedEnd",
                            "inputAdornedStart": "MuiInputBase-inputAdornedStart",
                            "inputHiddenLabel": "MuiInputBase-inputHiddenLabel",
                            "inputMarginDense": "MuiInputBase-inputMarginDense MuiInput-inputMarginDense",
                            "inputMultiline": "MuiInputBase-inputMultiline MuiInput-inputMultiline",
                            "inputSelect": "MuiInputBase-inputSelect",
                            "inputTypeSearch": "MuiInputBase-inputTypeSearch MuiInput-inputTypeSearch",
                            "marginDense": "MuiInputBase-marginDense",
                            "multiline": "MuiInputBase-multiline MuiInput-multiline",
                            "root": "MuiInputBase-root MuiInput-root MuiInput-underline",
                          }
                        }
                        fullWidth={false}
                        id="standard-dense"
                        inputComponent="input"
                        inputProps={
                          Object {
                            "data-test": "text-field",
                          }
                        }
                        multiline={false}
                        onChange={[Function]}
                        type="text"
                        value=""
                      >
                        <div
                          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
                          onClick={[Function]}
                        >
                          <input
                            aria-invalid={false}
                            autoComplete="off"
                            className="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
                            data-test="text-field"
                            disabled={false}
                            id="standard-dense"
                            onAnimationStart={[Function]}
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            required={true}
                            type="text"
                            value=""
                          />
                        </div>
                      </ForwardRef(InputBase)>
                    </WithStyles(ForwardRef(InputBase))>
                  </ForwardRef(Input)>
                </WithStyles(ForwardRef(Input))>
              </div>
            </ForwardRef(FormControl)>
          </WithStyles(ForwardRef(FormControl))>
        </ForwardRef(TextField)>
      </WithStyles(ForwardRef(TextField))>
    </TextInput>
  </span>
  <WithStyles(ForwardRef(Fab))
    aria-label="add"
    className="makeStyles-fab-3"
    color="secondary"
    disabled={true}
    size="small"
    type="submit"
  >
    <ForwardRef(Fab)
      aria-label="add"
      className="makeStyles-fab-3"
      classes={
        Object {
          "colorInherit": "MuiFab-colorInherit",
          "disabled": "Mui-disabled",
          "extended": "MuiFab-extended",
          "focusVisible": "Mui-focusVisible",
          "label": "MuiFab-label",
          "primary": "MuiFab-primary",
          "root": "MuiFab-root",
          "secondary": "MuiFab-secondary",
          "sizeMedium": "MuiFab-sizeMedium",
          "sizeSmall": "MuiFab-sizeSmall",
        }
      }
      color="secondary"
      disabled={true}
      size="small"
      type="submit"
    >
      <WithStyles(ForwardRef(ButtonBase))
        aria-label="add"
        className="MuiFab-root makeStyles-fab-3 MuiFab-sizeSmall Mui-disabled MuiFab-secondary"
        component="button"
        disabled={true}
        focusRipple={true}
        focusVisibleClassName="Mui-focusVisible"
        type="submit"
      >
        <ForwardRef(ButtonBase)
          aria-label="add"
          className="MuiFab-root makeStyles-fab-3 MuiFab-sizeSmall Mui-disabled MuiFab-secondary"
          classes={
            Object {
              "disabled": "Mui-disabled",
              "focusVisible": "Mui-focusVisible",
              "root": "MuiButtonBase-root",
            }
          }
          component="button"
          disabled={true}
          focusRipple={true}
          focusVisibleClassName="Mui-focusVisible"
          type="submit"
        >
          <button
            aria-label="add"
            className="MuiButtonBase-root MuiFab-root makeStyles-fab-3 MuiFab-sizeSmall Mui-disabled MuiFab-secondary Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="submit"
          >
            <span
              className="MuiFab-label"
            >
              <ForwardRef>
                <WithStyles(ForwardRef(SvgIcon))>
                  <ForwardRef(SvgIcon)
                    classes={
                      Object {
                        "colorAction": "MuiSvgIcon-colorAction",
                        "colorDisabled": "MuiSvgIcon-colorDisabled",
                        "colorError": "MuiSvgIcon-colorError",
                        "colorPrimary": "MuiSvgIcon-colorPrimary",
                        "colorSecondary": "MuiSvgIcon-colorSecondary",
                        "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                        "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                        "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                        "root": "MuiSvgIcon-root",
                      }
                    }
                  >
                    <svg
                      aria-hidden="true"
                      className="MuiSvgIcon-root"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                    </svg>
                  </ForwardRef(SvgIcon)>
                </WithStyles(ForwardRef(SvgIcon))>
              </ForwardRef>
            </span>
          </button>
        </ForwardRef(ButtonBase)>
      </WithStyles(ForwardRef(ButtonBase))>
    </ForwardRef(Fab)>
  </WithStyles(ForwardRef(Fab))>
</form>
`;
